package/MDAnalysis/analysis/hbonds/wbridge_analysis.py

```diff
@@ -987,29 +991,48 @@ def _prepare(self):
 
-    def _donor2acceptor(self, donors, donor_hs, acceptor):
+    def _donor2acceptor(self, donors, h_donors, acceptor):
+        if len(donors) == 0 or len(acceptor) == 0:
+            return []
+        if self.distance_type != 'heavy':
+            donors_idx = list(h_donors.keys())
+        else:
+            donors_idx = list(donors.keys())
         result = []
-        ns_acceptors = AtomNeighborSearch(acceptor, self.box)
-        for i, donor_h_set in donor_hs.items():
-            d = donors[i]
-            for h in donor_h_set:
-                res = ns_acceptors.search(h, self.distance)
-                for a in res:
-                    donor_atom = h if self.distance_type != 'heavy' else d
-                    dist = distances.calc_bonds(donor_atom.position,
-                                                a.position, box=self.box)
-                    if dist <= self.distance:
-                        angle = distances.calc_angles(d.position, h.position,
-                                                      a.position, box=self.box)
-                        angle = np.rad2deg(angle)
-                        if angle >= self.angle:
-                            self.logger_debug(
-                                "D: {0!s} <-> A: {1!s} {2:f} A, {3:f} DEG" \
-                                    .format(h.index, a.index, dist, angle))
-                            result.append((h.index, d.index, a.index,
-                                     (h.resname, h.resid, h.name),
-                                     (a.resname, a.resid, a.name),
-                                     dist, angle))
+        # Code modified from p-j-smith
+        pairs, distances = capped_distance(self.u.atoms[donors_idx].positions,
+                                           self.u.atoms[acceptor].positions,
+                                           max_cutoff=self.distance, box=self.box,
+                                           return_distances=True)
+        if self.distance_type != 'heavy':
+            tmp_donors = [h_donors[donors_idx[idx]] for idx in pairs[:,0]]
+            tmp_hydrogens = [donors_idx[idx] for idx in pairs[:,0]]
+            tmp_acceptors = [acceptor[idx] for idx in pairs[:,1]]
+        else:
+            tmp_donors = []
+            tmp_hydrogens = []
+            tmp_acceptors = []
+            for idx in range(len(pairs[:,0])):
+                for h in donors[donors_idx[pairs[idx,0]]]:
+                    tmp_donors.append(donors_idx[pairs[idx,0]])
+                    tmp_hydrogens.append(h)
+                    tmp_acceptors.append(acceptor[pairs[idx,1]])
+
+        angles = np.rad2deg(
+            calc_angles(
+                self.u.atoms[tmp_donors].positions,
+                self.u.atoms[tmp_hydrogens].positions,
+                self.u.atoms[tmp_acceptors].positions,
+                box=self.box
+            )
+        )
+        hbond_indices = np.where(angles > self.angle)[0]
+        for index in hbond_indices:
```

Comment:
Thank you for the suggestion. I will make it into a PR after this one is merged.

Reply:
After given it some thought, I'm not quite sure of the representation that you are talking about.
I will start with a list of indices (hydrogen bond donor heavy atom, hydrogen bond donor hydrogen, hydrogen bond acceptor), which is an array of rough size of (1000,3).
Then I will do the distance search, which gives the distance and shrink the number of indexes significantly. Thus, we have an array of (100,4), where the last column is distance.
Then I will do the angle search, which will also shrink the number of indexes and gives the final result of (10,5), where the last column is the angle.

To make a structured array, we need to know the array size in advance.
Do you suggest that I shall make a structured array of (1000,5) and then slice through it? to eventually get to (10,5).
Or is it better to reconstructed an array ar each step?
(1000,3) > (100,4) > (10,5)
Thank you.

Reply:
You know the size of hbond_indices, so rather than have a loop iteration for each entry, you could create a structured array of size hbond_indices and fill that. Not too important though